contracts/src/OffChainMessageContracts/ValidatorSetSig.sol

```diff
@@ -0,0 +1,124 @@
+
+    /**
+     * @notice Address that the off-chain Warp message sets as the "source" address.
+     * @dev The address is not owned by any EOA or smart contract account, so it
+     * cannot possibly be the source address of any other Warp message emitted by the VM.
+     */
+    address public constant VALIDATORS_SOURCE_ADDRESS = address(0);
+
+    /**
+     * @notice The blockchain ID of the chain the contract is deployed on.
+     */
+    bytes32 public immutable blockchainID;
+
+    /**
+     * @notice Warp precompile used for sending and receiving Warp messages.
+     */
+    IWarpMessenger public constant WARP_MESSENGER =
+        IWarpMessenger(0x0200000000000000000000000000000000000005);
+
+    /**
+     * @notice Emited when the payload is successfully delivered to the target contract.
+     */
+    event Delivered(address indexed targetContractAddress, uint256 indexed nonce);
+
+    constructor(bytes32 validatorBlockChainID) {
+        validatorBlockchainID = validatorBlockChainID;
+        blockchainID = WARP_MESSENGER.getBlockchainID();
+    }
+
+    function executeCall(uint32 messageIndex) external nonReentrant {
+        // Get the WarpMessage from the WarpMessenger pre-compile and verify that it is valid
+        (WarpMessage memory message, bool valid) =
+            WARP_MESSENGER.getVerifiedWarpMessage(messageIndex);
+        require(valid, "ValidatorSetSig: invalid warp message");
+
+        // Ensure that the sourceChainID of the validator set that signed the message is the authorized one
+        require(
+            message.sourceChainID == validatorBlockchainID, "ValidatorSetSig: invalid sourceChainID"
+        );
+        // Ensure that the originSenderAddress is the zero address to prevent on-chain messages from being forwarded
+        require(
+            message.originSenderAddress == VALIDATORS_SOURCE_ADDRESS,
+            "ValidatorSetSig: non-zero originSenderAddress"
+        );
+
+        ValidatorSetSigMessage memory validatorSetSigMessage;
+        (validatorSetSigMessage) = abi.decode(message.payload, (ValidatorSetSigMessage));
+
+        bool success = validateMessage(validatorSetSigMessage);
+        require(success, "ValidatorSetSig: invalid message");
+
+        nonces[validatorSetSigMessage.targetContractAddress] = validatorSetSigMessage.nonce;
+
+        (success,) =
+        // solhint-disable-next-line avoid-low-level-calls
+         validatorSetSigMessage.targetContractAddress.call(validatorSetSigMessage.txPayload);
+
+        // Use require to revert the transaction if the call fails. This is to prevent consuming the nonce if the call fails due to OOG
+        // and requiring re-signing of the message with a new nonce.
+        require(success, "ValidatorSetSig: call failed");
+        emit Delivered(validatorSetSigMessage.targetContractAddress, validatorSetSigMessage.nonce);
+    }
+
+    function validateMessage(ValidatorSetSigMessage memory message) public view returns (bool) {
```

Comment:
If `validateMessage` returns a bool, I think the implementation should be along the lines of:
```
return 
    message.validatorSetSigAddress == address(this) && 
    message.targetBlockchain = blockchainID &&
    ...;
```

And then the handling of invalid messages (i.e. reverting) is left to the caller.

Alternatively, we could have the individual reverts here and not have any return value.

Reply:
Do you have a preference for either way? I agree that the current implementation is not very idiomatic. The main downside of switching to above implementation is that no specific reason for the revert is given to the caller. It's not a huge problem here since there are only a few fields so issues can be visually inspected.  

Overall slightly leaning towards keeping the reverts and removing the return value but don't feel strongly either way

Reply:
> Overall slightly leaning towards keeping the reverts and removing the return value but don't feel strongly either way

Yeah that makes sense to me